packages/react-compass/src/drawer/drawer.tsx

```diff
@@ -0,0 +1,189 @@
+import {pickChild} from '../utils/pick-child'
+import {StyledComponentProps} from '../utils/stitches.types'
+import {useDOMRef} from '../utils/use-dom-ref'
+import DrawerExpander from './drawer-expander'
+import DrawerFooter from './drawer-footer'
+import DrawerHeader from './drawer-header'
+import {
+  DrawerVariantProps,
+  StyledDrawer,
+  StyledDrawerBody,
+} from './drawer.styles'
+
+const bodyOverflowY =
+  typeof document !== 'undefined' ? document.body.style.overflowY : 'auto'
+
+type Props = (
+  | {
+      variant?: 'default'
+      position?: 'right' | 'bottom' | 'left' | undefined
+      onExpandChange?: never
+    }
+  | {
+      variant: 'h5'
+      position?: never
+      onExpandChange?: (isExpand: boolean) => void
+    }
+) &
+  StyledComponentProps
+
+export type DrawerProps = Props &
+  Omit<DrawerVariantProps, 'position'> &
+  Omit<DialogHTMLAttributes<HTMLDialogElement>, keyof Props>
+
+const Drawer = forwardRef<HTMLDialogElement, DrawerProps>((props, ref) => {
+  const {
+    className = '',
+    css = {},
+    style,
+    children,
+
+    open = false,
+    onClose,
+    onMouseDown,
+
+    variant = 'default',
+    position: drawerPosition = 'right',
+    onExpandChange,
+    ...delegated
+  } = props
+
+  const position: DrawerProps['position'] =
+    variant === 'h5' ? 'bottom' : drawerPosition
+  const DrawerRef = useDOMRef<HTMLDialogElement>(ref)
+  const DrawerElement = DrawerRef.current
+
+  const [drawerInitHeight, setDrawerInitHeight] = useState(0)
+  const [drawerHeight, setDrawerHeight] = useState(DrawerElement?.offsetHeight)
+  const [isExpanded, setIsExpanded] = useState(false)
+
+  const {
+    target: DrawerExpanderRef,
+    position: {y},
+    setPosition,
+  } = useDrag<HTMLDivElement>({
+    stepSize: 10,
+    direction: 'vertical',
+    setCSS: false,
+    onEnd(_, dragPosition, setPositionOnEnd) {
+      if (!window.innerHeight || !open) {
+        return
+      }
+
+      const collapseHeight = Math.floor(window.innerHeight / 3)
+      let expandableLine = collapseHeight * 2
+      if (drawerInitHeight > expandableLine) {
+        expandableLine = drawerInitHeight
+      }
+
+      const newHeight = drawerInitHeight - dragPosition[1]
+
+      const isCrossExpandLine = newHeight > expandableLine
+      setIsExpanded(isCrossExpandLine)
+      onExpandChange?.(isCrossExpandLine)
+
+      // Close the drawer if the user drag the drawer to the bottom of screen
+      if (newHeight < 100) {
+        setDrawerHeight(drawerInitHeight)
+        DrawerElement?.close()
+        return
+      }
+
+      setPositionOnEnd([0, 0])
+    },
+  })
+
+  const {child: DrawerHeaderElement, rest: OtherElementsExceptHeader} =
+    pickChild<typeof DrawerHeader>(children, DrawerHeader)
+
+  const {child: DrawerFooterElement, rest: OtherElements} = pickChild<
+    typeof DrawerFooter
+  >(OtherElementsExceptHeader, DrawerFooter)
+
+  const handleMouseDown = (e: MouseEvent<HTMLDialogElement>) => {
```

Comment:
using `useCallback`

Reply:
no need to, this function will be fired only once when the user clicks on the Drawer.

Reply:
if `onMouseDown` props change, this function will be re-rendered.

Reply:
updated